### ISSUE TITLE
Code Folding with massive files locks UI

### ISSUE THREAD

**Issue by AmericanKulak** [codemirror/CodeMirror#4504]
When using code folding with massive documents (Extremely noticeable at ~25000 lines):

Collapsing the entire document calls 'onViewportChange', where the viewport.to property reports the end of the document (~25000). 

This, in turn, calls updateFoldInfo with a range that is essentially the entire document. 

updateFoldInfo then ranges over the entire range, checking the fold gutter mark and updating it for each line.

This locks the eventLoop until the entire range (which is invisible due to folding anyway) is updated.

**Comment by adrianheine** [comment]
@marijnh I was thinking we could skip `setGutterMark` for all folded lines:

```javascript
  function updateFoldInfo(cm, from, to) {
    var opts = cm.state.foldGutter.options, cur = from;
    var minSize = cm.foldOption(opts, "minFoldSize");
    var func = cm.foldOption(opts, "rangeFinder");
+   var done = false;
    cm.eachLine(from, to, function(line) {
+     if (done) return;
      var mark = null; 
      if (isFolded(cm, cur)) {
        mark = marker(opts.indicatorFolded);
+       done = true;
      } else {
        var pos = Pos(cur, 0);
        var range = func && func(cm, pos);
        if (range && range.to.line - range.from.line >= minSize)
          mark = marker(opts.indicatorOpen);
      }
      cm.setGutterMarker(line, opts.gutter, mark);
      ++cur;
    });
```

This is obviously too naive, but `isFolded` returns the fold mark, and I was hoping from that we can determine which lines are folded?

**Comment by marijnh** [comment]
> I was thinking we could skip setGutterMark for all folded lines

That's doable, but does it help? I can imagine that a lot of the time is lost in the actual fold-finding function, which will still run over (and usually tokenize) the whole folded blob.

**Comment by vincenttang96** [comment]
any updates on this issue? I'm still having this issue in v5.55.0.

**Comment by marijnh** [comment]
I don't think this is likely to be fixed in version 5. (The rewrite for version 6 models folding in a more efficient way, and shouldn't have this issue.)
